### ISSUE TITLE
SignalPlot fails when arrays contain double.NaN

### ISSUE THREAD

**Issue by bukkideme** [swharden/ScottPlot#546]
**Describe the bug**
When I create a second SignalPlot using a List, it is only shown after zoom in. I guess I only miss something basic...?

**To Reproduce**
Create a SignalPlot with rand numbers, then a List based on this array, and send it to a second SignalPlot.

```cs
// include code to help us reproduce this bug
using MathNet.Numerics.Statistics;
using ScottPlot;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace ScottPlot_PlotSignalNotShown
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            Random rand = new Random(0);
            int pointCount = (int)1e5;
            double[] array1 = DataGen.RandomWalk(rand, pointCount);
            
            formsPlot1.plt.PlotSignal(array1, 1, 0, 0, color: Color.Blue, 2, 6, "Signal 1");
            
            List<double> smooth = new List<double>();
            MovingStatistics moving_aver = new MovingStatistics(24);
            foreach (double value in array1)
            {
                moving_aver.Push(value);
                smooth.Add(Math.Round(moving_aver.Variance));
            }

            formsPlot1.plt.PlotSignal(smooth.ToArray(), 1, 0, 0, color: Color.Green, 2, 6, "Signal 1 variance");

            formsPlot1.plt.Legend(true, fontSize: 18, location: legendLocation.upperRight);
            formsPlot1.Render();
        }
    }
}
```

**Screenshots**
![image](https://user-images.githubusercontent.com/51174971/93259334-004b8a00-f7a0-11ea-91a7-b7e5ba807164.png)

After zoom in:

![image](https://user-images.githubusercontent.com/51174971/93259362-093c5b80-f7a0-11ea-91a4-4307c43e83fb.png)

**Comment by StendProg** [comment]
Hi @bukkideme,
Probably `smooth` array contains `NAN` or `Infinity` values. Check for this especially on borders.
All others looks fine.

**Comment by bukkideme** [comment]
Ups! Thanks, i totally forgot that the variance calculation needs at least 2 values, it returns NaN for the first iteration at the single element :)
Is there any plan to make a ScottPlot plot give a warning in such case? Like when a nan value is given by mistake...
Thanks!

**Comment by StendProg** [comment]
Personally I dislike this types of checks. Full array enumeration requires for this. Millions of points is the typical use case for `PlottableSignal`. Throw exception is the only one way to warning user, it also not so good solution. There is possible to make helper method to allow user to check his data before plotting. But let's be honest, nobody will use it.

**Comment by swharden** [comment]
Hi @bukkideme, it seems like your issue has been solved!

> Is there any plan to make a ScottPlot plot give a warning in such case? Like when a nan value is given by mistake...
Thanks!

I agree with @StendProg that such a check (iterating through every point of large arrays) is very slow, and not good to do automatically on every render.

However, it may be useful to add a method like `bool IsValidData()` so the user can decide if/when to check data for validity. I'll consider adding something like this to the Plottable interface in ScottPlot5. Thanks for the suggestion!

**Comment by StendProg** [comment]
After some thought, I found that it would be handy. When the user is faced with an incomprehensible problem, give him the opportunity to turn on the `diagnostic mode` in one place at init. Something like: 
```c#
plt.EnableDiagnostic ().
```
As a result, all data would be checked on every Render, and throw exceptions describing the problem. Performance is irrelevant in diagnostic mode. I still vaguely imagine how this can be done through decorators, so that it absolutely does not affect the standard mode, but I will see what can be done.
In the future, this mode can be expanded with additional checks.

**Comment by swharden** [comment]
I like this line of thinking. If there is a method that can check the validity of the data, it would be nice if the user can run that method once, or enable it to be run before every render. A command that controls this behavior for all plottables seems useful.  Perhaps something like this?

```cs
var scatter = new plt.PlotScatter(xs, ys);
scatter.ValidateData(); // to do it just once
scatter.ValidateDataBeforeEveryRender = true; // diagnostic mode
```

... and perhaps plot-level code to do a similar thing for all plottables like:
```cs
plt.ValidateData(); // to do it just once
plt.ValidateDataBeforeEveryRender = true; // diagnostic mode
```

**Comment by StendProg** [comment]
I am thinking of a slightly different design.
In Diagnostic mode, before adding a plottable to list, wrap it in `CheckDecorators`  and add them to list. Each Decorator will call the appropriate check before delegating the Render call.
```c#
public class CheckForNANDecorator() : IPlottable
{
  private IPlottable plottable;

  public CheckForNANDecorator(IPlottable plottable)
  {
    this.plottable = plottable
  }

  public void Render(Settings settings)
  {
    CheckForNAN();
    plottable.Render(settings);
  }
}

///---
PlotSignal(...)
{
  var signal = new PlottableSignal(...);
  IPlottable wrappedPlottable = signal;
  if (diagnostic)
  {
    wrappedPlottable = WrapWithDiagnostic(wrappedPlottable);    
  }
 plot.Add(wrappedPlottable)
 return signal;
}

public IPlottable WrapWithDiagnostic(IPlottable wrappedPlottable)
{
  if (diagnosticSettings.CheckForNan)
    {
      wrappedPlottable = new CheckForNANDecorator(wrappedPlottable)
    } 
   if (otherCheck)
   {
      wrappedPlottable = new OtherCheckDecorator(wrappedPlottable);
   }
   .....
}
```
In this case, in standard mode, only one `if` will be added in `plottable.Add` section, which is not essential for performance.
The advantage of this approach is that it will be very easy to add new checks.
drawbacks, the Plottable list will no longer contain specific plotables, and probably a lot more, but I don't see it to the end yet.

**Comment by StendProg** [comment]
> In Diagnostic mode, before adding a plottable to list, wrap it in CheckDecorators and add them to list. Each Decorator will call the appropriate check before delegating the Render call.

I prototyped this, and it works fine. I'll make a PR in a few days.

Some problems to discuss:
- `Plot` have `Add(plottable)` method, but all `PlotSignal, PlotScatter, ...` use directly `settings.plottable.Add()`. Is this acceptable to replace them with `Plot.Add()`.

- There is a lot of `Plottables` and most of them different. To understand something more about them, for example, whether they have an `ys` array, need to provide additional information.  This can be solved by one of this solution:
1. All plottable that have `ys` must implement `IYData`. But `ys` not property for now.
2. Mark plottable wiith `[AllowYDataCheck]` attribute. This simplifies the choice of a decorator, but the data itself will need to be received through reflection.
3. It is possible  check for the presence of the required field and pick up the data using reflection. Only low performance is a disadvantage and this is not important for `Diagnostic mode`. I have stopped at this approach for now.
